src/mjolnir/landmark_database_builder.cc

```diff
@@ -1,103 +1,114 @@
 #include "mjolnir/landmark_database_builder.h"
+#include "filesystem.h"
 #include "mjolnir/util.h"
 
 namespace valhalla {
 namespace mjolnir {
-void LandmarkDatabase::connect_database() {
-  if (!open_database()) {
-    throw std::runtime_error("Cannot open database");
-    return;
-  }
 
-  if (open_flags == (SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE)) {
-    if (!create_landmarks_table() || !create_spatial_index()) {
-      throw std::runtime_error("Cannot create landmarks table or spatial index");
-      return;
+// TODO: this can be a utility and be more generic with a few more options, we could make the prepared
+//  statements on the fly and retrievable by the caller, then anything in the code base that wants to
+//  use sqlite can make use of this utility class. for now its ok to be specific to landmarks though
+struct LandmarkDatabase::db_pimpl {
+  sqlite3* db;
+  sqlite3_stmt* insert_stmt;
+  sqlite3_stmt* bounding_box_stmt;
+  std::shared_ptr<void> spatial_lite;
+  bool vacuum_analyze = false;
+
+  db_pimpl(const std::string& db_name, bool read_only) : insert_stmt(nullptr) {
```

Comment:
the constructor does all the initialization stuff directly rather than having myriad functions. if anything goes wrong it throws. if something upstream wants to catch what is thrown and log it fine. i would expect the graph enhancer for example to catch, log that i couldnt open the landmarks database and that its skipping adding landmarks to the tiles. for the writer part i would just let the throw abort the process (ie dont catch)